### PR TITLE
Project invitation emails also register new users

### LINKED ISSUE #239: invite users to join LD when invited as a project member
right when a user adds a project member, if there's no user matching that email address then nothing happens. We would like to send that user an email so they can create an account.

### PR DESCRIPTION
**Pull request by rmunn**
Project managers who want to invite new users can click one button and send an invitation email.

We could almost reuse the "Verify email" flow, except for one problem: new users won't have a password. We could send them to the forgot-password page next, but that's not great UX. Better to create a new flow that they will land on, which asks them to choose a password then brings them to their home page.

Will fix #239 once completed.

### CODE REVIEW

**Comment by rmunn** [comment]
Design decision: the "You've been invited to join this project" email will link to the new user registration page, with the user's name and email address already filled in (and the JWT in a hidden form field). The password field will be left blank; once the user fills it in, the user will be registered. If the user did not change the email address in the registration form, he/she can skip verifying the email. But we also allow changing the email address, in which case the user will have to go through the verification process.

**Comment by rmunn** [comment]
Email registration flow is working now, but I want to tweak three things in the UI:

- There needs to be feedback from the "Add Member" modal dialog that says "Email invitation sent; user will be added once they click on the invite link". Right now it shows the "User has not verified their email address" error because I reused that particular exception as a placeholder. I need to create a new exception to return to the frontend UI, and code in the frontend to handle that by displaying a notification and dismissing the dialog.
- When the user receives the invitation email, it takes them to the registration page so they can fill out their name and a password. That's fine. What is less fine is that when they finish the registration process, it takes them to `/home`. I'd like it to take them directly to the project they just signed up for, rather than their home page where there's only one thing to click on. Also, I'd like the registration page to look slightly different if you're registering as a project member, i.e. I'd like to have the project name visible somewhere on the registration page.
- In addition to showing the project name in the registration page, I'd like to show the project name, as well as the name of the manager sending the invitation, in the email. That means tweaking SendCreateAccountEmail to take two more parameters, inviter name and project name, and tweaking ProjectMutations.AddProjectMember to supply those parameters. (Both are easily looked up from the logged-in user ID and the project ID supplied in the input, so that's not hard).

Once I've made those three changes, this will be ready for review.

**Comment by rmunn** [comment]
Lint error is because I added an exception type to the AddProjectMember mutation in ProjectMutations.cs, but the GHA build isn't getting that new error type into the generated gql.ts file. So lint is saying "Hey, AddProjectMember.svelte is checking for a ProjectMemberInvitedByEmail error type that doesn't exist". It does, but the generated GQL Typescript is lagging behind reality in our GHA build, for a reason I haven't figured out yet.

**Comment by hahn-kev** [comment]
@rmunn you probably need to update the gql schema file. `task api:generate-gql-schema` should do that for you.

**Comment by rmunn** [comment]
> @rmunn you probably need to update the gql schema file. `task api:generate-gql-schema` should do that for you.

Thanks; I did that on the https://github.com/sillsdev/languageforge-lexbox/pull/571 PR but forgot I hadn't done it here.

**Review by myieye — CHANGES_REQUESTED** [review]

File: backend/LexBoxApi/Controllers/LoginController.cs [hunk 1]
```diff
@@ -89,6 +90,44 @@ private async Task<ActionResult> EmailLinkExpired()
...
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<LexAuthUser>> JoinProject(
+        string jwt, // This is required because auth looks for a jwt in the query string
+        string returnTo)
+    {
+        var userId = loggedInContext.User.Id;
+        // Ensure we're not registering a user ID twice
+        var user = await lexBoxDbContext.Users.FindAsync(userId);
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/LoginController.cs]: Made moot by commit 50d85fa, which removes this code.

File: backend/LexBoxApi/Controllers/LoginController.cs [hunk 2]
```diff
@@ -89,6 +90,44 @@ private async Task<ActionResult> EmailLinkExpired()
...
+            IsAdmin = false,
+            Locked = false,
+        };
+        if (loggedInContext.User.Projects.Length > 0)
+        {
+            user.Projects = loggedInContext.User.Projects.Select(p => new ProjectUsers { Role = p.Role, ProjectId = p.ProjectId }).ToList();
+        }
+
+        user.Email = loggedInContext.User.Email;
+        user.EmailVerified = true;
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/LoginController.cs]: Made moot by commit 50d85fa, which removes this code.

File: backend/LexBoxApi/Controllers/LoginController.cs [hunk 3]
```diff
@@ -89,6 +90,44 @@ private async Task<ActionResult> EmailLinkExpired()
...
+        var userId = loggedInContext.User.Id;
+        // Ensure we're not registering a user ID twice
+        var user = await lexBoxDbContext.Users.FindAsync(userId);
+        if (user != null) return Unauthorized(); // TODO: Maybe just add them to the project and move on?
+        user = new User
+        {
+            Id = userId,
+            Name = "",
+            Email = loggedInContext.User.Email,
+            PasswordHash = "", // New users get sent to forgot-password page next
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/LoginController.cs]: Made moot by commit 50d85fa, which removes this code.

File: frontend/src/routes/(unauthenticated)/register/+page.svelte [hunk 4]
```diff
@@ -4,8 +4,15 @@
...
   import { register } from '$lib/user';
+  import { getSearchParamValues } from '$lib/util/query-params';
+  import { onMount } from 'svelte';
   import { z } from 'zod';
 
+  type RegisterPageQueryParams = {
+    name: string;
+    email: string;
+    locale: string;
+  };
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/routes/(unauthenticated)/register/+page.svelte:15]: I agree on `locale`, but I do want to be able to support `name`. We'd have to modify the UI in the "Add member" modal to allow specifying the name of the user we want to invite, as well as their email address. We might even look for email address input in the form `Real Name <email@domain>` and parse it, as a bells-and-whistles kind of feature, so that if someone pastes in some address from their address book, we can extract the name from it.

Also, if the "Add Member" UI allowed managers to specify the name of the user they're adding, I could use that in the email template. Right now the email template is the only one that *doesn't* greet the recipient by name, and that feels weird.

I'll remove `locale` for now, and see whether we should remove `name` as well or whether we want to modify the "Add Member" UI to allow specifying the name of the member we're adding.

File: backend/LexBoxApi/Services/EmailService.cs [hunk 5]
```diff
@@ -148,4 +187,11 @@ private static MimeMessage StartUserEmail(User user, string? email = null)
         message.To.Add(new MailboxAddress(user.Name, email ?? user.Email));
         return message;
     }
+
+    private static MimeMessage StartUserEmail(string name, string email)
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Services/EmailService.cs:190]: Done in commit 83144638.

File: backend/LexBoxApi/Controllers/UserController.cs [hunk 6]
```diff
@@ -50,13 +50,16 @@ public async Task<ActionResult<LexAuthUser>> RegisterAccount(RegisterAccountInpu
     {
         using var registerActivity = LexBoxActivitySource.Get().StartActivity("Register");
         var validToken = await _turnstileService.IsTokenValid(accountInput.TurnstileToken, accountInput.Email);
+        var jwtUser = _loggedInContext.MaybeUser;
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/UserController.cs:53]: Done in commit 83144638.

File: backend/LexBoxApi/Controllers/UserController.cs [hunk 7]
```diff
@@ -50,13 +50,16 @@ public async Task<ActionResult<LexAuthUser>> RegisterAccount(RegisterAccountInpu
...
         registerActivity?.AddTag("app.turnstile_token_valid", validToken);
         if (!validToken)
         {
             ModelState.AddModelError<RegisterAccountInput>(r => r.TurnstileToken, "token invalid");
             return ValidationProblem(ModelState);
         }
 
+        var emailVerified = jwtUser?.Email == accountInput.Email;
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/UserController.cs:61]: Done in commit 83144638.

File: frontend/src/routes/(unauthenticated)/register/+page.svelte [hunk 8]
```diff
@@ -34,6 +41,15 @@
...
     throw new Error('Unknown error, no error from server, but also no user.');
   });
+  onMount(() => { // query params not available during SSR
+    const urlValues = getSearchParamValues<RegisterPageQueryParams>();
+    form.update((form) => {
+      if (urlValues.name) form.name = urlValues.name;
+      if (urlValues.email) form.email = urlValues.email;
+      if (urlValues.locale) form.locale = urlValues.locale;
+      return form;
+    }, { taint: false });
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/routes/(unauthenticated)/register/+page.svelte]: Done in commit 83144638.

File: frontend/src/routes/(unauthenticated)/register/+page.svelte (hunk 4 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/routes/(unauthenticated)/register/+page.svelte]: Removed `locale` in commit 83144638. Let's discuss whether adding "Name" to the "Add Member" UI is worth doing, and how, before I resolve this conversation.

[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/routes/(unauthenticated)/register/+page.svelte]: Removed `name` in commit eec5af65.

File: backend/LexBoxApi/Controllers/LoginController.cs [hunk 9]
```diff
@@ -41,10 +42,14 @@ public class LoginController(
         string returnTo)
     {
         var user = loggedInContext.User;
-        var userUpdatedDate = await userService.GetUserUpdatedDate(user.Id);
-        if (userUpdatedDate != user.UpdatedDate)
+        // A RegisterAccount token means there's no user account yet, so checking UpdatedDate makes no sense
```
Comment (myieye) [inline backend/LexBoxApi/Controllers/LoginController.cs:45]: I still think it would be nice if we could handle a user getting multiple invites.
I think it makes sense to do something like this:
- If it's a `RegisterAccount` audience, check if an account with the email address already exists. If so:
  - Add any projects from this token to that account
  - Give them a new up-to-date token
  - Make the register page smarter (it's sort of a bug that an authenticated user can access the register page right now): if the user is authenticated (and not with `RegisterAccount`) redirect them home

**Review by myieye — CHANGES_REQUESTED** [review]
A few more things. The biggest one is being able to handle multiple invite emails.

File: backend/LexBoxApi/Controllers/LoginController.cs (hunk 9 above)
Comment (rmunn) [reply backend/LexBoxApi/Controllers/LoginController.cs:45]: I don't think it's realistic that users would be receiving invites to multiple projects at the same time. Our typical user who doesn't have an account yet is a mother-tongue speaker who's being asked to join a language project to contribute his vocabulary/grammar knowledge. Anyone who might be asked to join multiple projects is almost certainly a language consultant, who almost certainly has a Language Depot account already.

I do agree that the registration page could be smarter: now that we have a RegisterAccount audience, only accept a token with that audience and redirect any other tokens home. But I don't think allowing multiple project invitations will be worth the programmer effort involved.

Comment (hahn-kev) [reply backend/LexBoxApi/Controllers/LoginController.cs:45]: created #584

File: frontend/src/lib/i18n/locales/en.json [hunk 10]
```diff
@@ -216,7 +217,8 @@ the [Linguistics Institute at Payap University](https://li.payap.ac.th/) in Chia
       "user_delete": "{name} has been removed.",
       "rename_project": "Project name set to {name}.",
       "describe": "Project description has been updated.",
-      "add_member": "{email} has been added to project."
+      "add_member": "{email} has been added to project.",
+      "member_invited": "{email} has been sent an invitation email to join the project."
```
[the remark this answers is not in the document]
Comment (rmunn) [reply frontend/src/lib/i18n/locales/en.json:221]: I tried to write something like that and it came out too wordy, so I cut it. Yours is much better. Done.

File: backend/LexBoxApi/Controllers/UserController.cs [hunk 11]
```diff
@@ -68,26 +71,30 @@ public async Task<ActionResult<LexAuthUser>> RegisterAccount(RegisterAccountInpu
         var salt = Convert.ToHexString(RandomNumberGenerator.GetBytes(SHA1.HashSizeInBytes));
         var userEntity = new User
         {
-            Id = Guid.NewGuid(),
+            Id = jwtUser?.Id ?? Guid.NewGuid(),
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/UserController.cs:74]: Whoops, I made the edits but didn't hit save, so they didn't make it into the commit I pushed. Done in commit c81e615f.

[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/UserController.cs:74]: Except for the `jwtUser?.Id` thing, which is there so you can't accidentally click on the same JWT link twice and create two accounts. The GUID from the JWT (if there is one) will be used to register the user account, so if you click on it a second time you'll be told that the user account already exists. At least, that's the idea here. Maybe it's not needed, though.

File: backend/LexBoxApi/Services/Email/EmailTemplates.cs [hunk 12]
```diff
@@ -18,13 +18,17 @@ public enum EmailTemplate
...
     PasswordChanged,
+    CreateAccountRequest,
     CreateProjectRequest
 }
 
 public record ForgotPasswordEmail(string Name, string ResetUrl) : EmailTemplateBase(EmailTemplate.ForgotPassword);
 
 public record VerifyAddressEmail(string Name, string VerifyUrl, bool newAddress) : EmailTemplateBase(EmailTemplate.VerifyEmailAddress);
 
+// TODO: For better email body, will want project name here too, and maybe name of person inviting the user?
```
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Services/Email/EmailTemplates.cs]: Done.

File: backend/LexBoxApi/Controllers/UserController.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/UserController.cs:74]: Yeah, it's not needed since we are checking `hasExistingUser` by email address further up. I'll get rid of it.

[the remark this answers is not in the document]
Comment (rmunn) [reply backend/LexBoxApi/Controllers/UserController.cs]: GUID switch is in commit 2a241a58.

File: backend/LexBoxApi/Controllers/LoginController.cs (hunk 9 above)
Comment (rmunn) [reply backend/LexBoxApi/Controllers/LoginController.cs:45]: Since we're now tracking this idea in #584, I'll resolve this conversation.

**Comment by rmunn** [comment]
@myieye I've implemented your suggestions, and punted the "invite to multiple projects at once" idea to #584 to work on in a separate PR. Is this ready to merge now, or is there anything else you want to address while this PR is open?

**Review by myieye — APPROVED** [review]
Cool beans 😎
